lib/browser/api/web-contents.ts

```diff
@@ -729,6 +729,56 @@ WebContents.prototype._init = function () {
     }
   });
 
+  const originCounts = new Map<string, number>();
+  const openDialogs = new Set<AbortController>();
+  this.on('-run-dialog' as any, async (info: {frame: WebFrameMain, dialogType: 'prompt' | 'confirm' | 'alert', messageText: string, defaultPromptText: string}, callback: (success: boolean, user_input: string) => void) => {
```

Comment:
This isn't specific to this PR as we follow this `-*` event style elsewhere, but to avoid folks hacking things into our internals can we make a second event bus for these internals? Either a private member or hidden behind a symbol so that we can do this implementation without other folks reaching in and messing with it

Reply:
yeah, i've thought about that in the past too. In this particular case this `openDialogs` set is needed to prevent segfaults (as calling the callback after the dialog has been canceled = 💥), which is a bit janky but I didn't feel like writing the tracking code in C++. But especially for more delicate things like this it should probably be a bit harder to mess with.

That said: what's the real problem with letting people mess with this? It could theoretically let them/us fix or workaround issues without having to patch or update Electron. It's dangerous, but flexible.

Reply:
> That said: what's the real problem with letting people mess with this? It could theoretically let them/us fix or workaround issues without having to patch or update Electron. It's dangerous, but flexible.

* We don't want end users shooting themselves in the foot, touching / relying on this internal stuff could easily break in future versions with little to no warning.  Ref: `-new-window` private event for instance and the pain that caused even us.
* People meddling with internals causes all kinds of hard to diagnose issues because they very rarely even raise issues with that critical piece of information, it's why node takes a strong stance of hiding internals nowadays.  Ref: the graceful-fs days of old
* There's no reason for folks to be messing with these internals, if there's a bug, we should fix it, if they want to do something we don't want to support then there's probably a good reason and again, we don't want them doing it.

Yes, in theory, there's one person out there who will have a valid usecase for messing with our internals.  But for every one of them there's a thousand other devs who will hack away until their thing works with no regard for the future consequences.

This is the same logic as to why we separated internal / external IPC buses, why we hide certain fields, methods, etc. from the public API.  Why we use `js2c` and the internal module loader instead of nodes public module loader.  Everything that we expose, whether we like it or not someone will use, if we don't expose it, they can't use it 👍

Reply:
I mean, sort of true, it depends on how useful the exposed thing is :P

It's a trade-off. It's extra work for us to hide this stuff, but it's also extra work to handle support/issues from people holding the thing wrong. Which is greater? It's hard to know for sure. `-new-window` was supporting a lot of very common use cases. I think this one is a lot rarer/less useful.

In sum, if this private C++ -> JS bridge was already common I'd use it, but I don't think it's worth inventing for this.